src/coreclr/jit/optimizer.cpp

```diff
@@ -4916,13 +4397,27 @@ void Compiler::optFindNewLoops()
 {
     m_loops = FlowGraphNaturalLoops::Find(m_dfsTree);
 
-    if (optCanonicalizeLoops(m_loops))
+    bool modified = false;
+
+    if (optCompactLoops())
+    {
+        fgInvalidateDfsTree();
+        m_dfsTree = fgComputeDfs();
+        m_loops   = FlowGraphNaturalLoops::Find(m_dfsTree);
+        modified  = true;
+    }
```

Comment:
The only reason this is needed is because `optCompactLoops` needs to insert blocks to fix up fallthrough. Otherwise the DFS tree would not be invalidated by it. It would be possible to avoid this by waiting with fixing up the fall through until after we have also canonicalized the loops, e.g. by having some `fgFixFallthrough` like we discussed the other day.